cpp/distributed/dist-mnist.cpp

```diff
@@ -1,19 +1,11 @@
+#include <c10d/ProcessGroupMPI.hpp>
```

Comment:
this line leads to error!

Reply:
Maybe is it related to the fact that the default PyTorch installation excludes support for MPI? It says that you need to recompile PyTorch from scratch if you want to use the MPI backend for torch.distributed. (https://pytorch.org/tutorials/intermediate/dist_tuto.html#communication-backends) As I'm also having this exact same issue, I'll try to do that tomorrow.

Reply:
@osalpekar How can this be fixed? I am not sure if @lasagnaphil ' s idea will work since Libtorch doesn't need to be compiled?

Reply:
The libtorch installation files that pytorch.org distributes are all precompiled libraries (with their accompanying headers). Chances are that even if you use the system's package manager (such as the Arch Linux AUR), they will use those official precompiled libraries. And from what I've seen, these do not include any support for MPI (for example, it's missing the header ProcessGroupMPI.hpp)

It seems that the build servers for pytorch.org do not include MPI support, which is understandable because there are multiple MPI implementations (OpenMPI, MPICH, MVAPICH, ...) and it's impractical to build and distribute the library separately for each library. So the only option would be to build and install libtorch from scratch. Also, the CMake script for c10d seems to include MPI support for the library only if a valid MPI installation is found (https://github.com/pytorch/pytorch/blob/c215ffb6a2de1c85a62a146a9e282908335ba3ed/torch/lib/c10d/CMakeLists.txt#L146), so you need to make sure an MPI installation exists on your system before calling ``python setup.py install``.

Reply:
As a side note, this might indicate that NCCL would make for a simpler example, though certainly less interesting in terms of showing how to deal with some more internals (e.g. intialization, ...).

Reply:
@lasagnaphil I see, thanks! Were you able to compile libtorch from scratch using any MPI?

Reply:
Compiling works now, as the headers (at least) are now properly installed on the system. However, I'm still having a linking error (``undefined reference to `c10d::ProcessGroupMPI::createProcessGroupMPI``) that I need to investigate. I tried to link the ``c10d`` CMake target to my executable but it says that it doesn't exist, which is a mystery.

Reply:
Ah, I figured it out. It seems like the file ``libc10d.a`` wasn't copied properly into the installation folder. Also, it seems that the particular static library needs to be linked directly with the exact path as the following

```
target_link_libraries(dist-mnist ${CMAKE_PREFIX_PATH}/lib/libc10d.a)
```
After this, and fixing a couple of more errors, I was able to successfully run the example (training seems to be doing all fine). Here's the pull request for the fix: https://github.com/soumyadipghosh/examples/pull/1

Reply:
@lasagnaphil Great! Can you share how you compiled libtorch from scratch using MPI?

Reply:
It took a while to figure out how to compile libtorch cleanly, I'll provide you a shell script: https://gist.github.com/lasagnaphil/3e0099816837318e8e8bcab7edcfd5d9

There are a lot of flags, and you might need to tweak some of them depending on your PC / usecase.

Take in mind that it takes quite some CPU time to compile! (For my Ryzen 7 2700X system, it took about 20~30 minutes)

Reply:
@soumyadipghosh The PR should be in good shape after @lasagnaphil's patch - we should be good to go ahead and approve. Are you still seeing build issues?

Reply:
@osalpekar Sorry, I didn't have the chance to verify - will do it soon. For now, I have merged @lasagnaphil 's patch and added the link to the shell script for build in the Readme. We should be good to go ahead now. If I have build issues later, I will create an issue.

Reply:
@soumyadipghosh No problem, I'm accepting this for now. Would be great if we could verify the build before merging